stdlib/public/SDK/Foundation/Data.swift

```diff
@@ -844,221 +541,1334 @@ internal final class _DataStorage {
+            precondition(range.lowerBound <= endIndex, "index \(range.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            ensureUniqueReference()
+            storage.resetBytes(in: range)
+            if slice.upperBound < range.upperBound {
+                slice = slice.lowerBound..<HalfInt(range.upperBound)
+            }
+        }
+
+        @inlinable
+        mutating func replaceSubrange(_ subrange: Range<Index>, with bytes: UnsafeRawPointer?, count cnt: Int) {
+            precondition(startIndex <= subrange.lowerBound, "index \(subrange.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(subrange.lowerBound <= endIndex, "index \(subrange.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(startIndex <= subrange.upperBound, "index \(subrange.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(subrange.upperBound <= endIndex, "index \(subrange.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+
+            let nsRange = NSRange(location: subrange.lowerBound, length: subrange.upperBound - subrange.lowerBound)
+            ensureUniqueReference()
+            let upper = range.upperBound
+            storage.replaceBytes(in: nsRange, with: bytes, length: cnt)
+            let resultingUpper = upper - nsRange.length + cnt
+            slice = slice.lowerBound..<HalfInt(resultingUpper)
+        }
+
+        @inlinable
+        func copyBytes(to pointer: UnsafeMutableRawPointer, from range: Range<Int>) {
+            precondition(startIndex <= range.lowerBound, "index \(range.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(range.lowerBound <= endIndex, "index \(range.lowerBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(startIndex <= range.upperBound, "index \(range.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            precondition(range.upperBound <= endIndex, "index \(range.upperBound) is out of bounds of \(startIndex)..<\(endIndex)")
+            storage.copyBytes(to: pointer, from: range)
+        }
+
+        @inlinable
+        var hashValue: Int {
+            let hashRange = startIndex..<Swift.min(startIndex + 80, endIndex)
+            return storage.withUnsafeBytes(in: hashRange) {
+                return Int(bitPattern: CFHashBytes(UnsafeMutablePointer(mutating: $0.baseAddress?.assumingMemoryBound(to: UInt8.self)), $0.count))
+            }
+        }
+    }
+
+    @usableFromInline
+    @_fixed_layout
+    internal final class RangeReference {
+        @usableFromInline
+        var range: Range<Int>
+
+        @inlinable
+        var lowerBound: Int { return range.lowerBound }
+
+        @inlinable
+        var upperBound: Int { return range.upperBound }
+
+        @inlinable
+        var count: Int { return range.upperBound - range.lowerBound }
```

Comment:
Re: this and `range.isEmpty` — we found during testing that because `range` does not implement `count` itself (but instead gets the implementation from a default implementation on `Collection`), calculating this ourselves was orders of magnitude faster than calling `range.count` (dispatching through to the protocol implementation added a lot of overhead), and this was making a big difference in our unit tests.

It does appear that `Range` implements `isEmpty` directly, so we should be able to fix `.isEmpty` above.

Reply:
Can you file a bug for the optimizer for that? We should try and ensure they're equivalent if possible. It might also be something that inlining too much is causing as the inliner might be giving up on a huge function that needs a specialized `count` the middle.